src/json-patch.ts

```diff
@@ -257,30 +270,120 @@ module jsonpatch {
+    if (path.indexOf('/') === -1 && path.indexOf('~') === -1) return path;
+    return path.replace(/~/g, '~0').replace(/\//g, '~1');
+  }
+  /**
+   * Unescapes a json pointer path
+   * @param path The escaped pointer
+   * @return The unescaped path
+   */
+  export function unEscapePath(path: string): string {
+    return path.replace(/~1/g, '/').replace(/~0/g, '~');
+  }
+
+  /**
+   * Retrieves a value from a JSON document by a JSON pointer.
+   * Returns the value.
+   *
+   * @param document The document to get the value from
+   * @param pointer an escaped JSON pointer
+   * @return The retrieved value
+   */
+  export function getValueByPointer(document, pointer) {
+    let pathSegments = [];
+    if (pointer) { //empty string evaluates to false
+      pathSegments = pointer.substring(1).split(/\//).map(jsonpatch.unEscapePath);
+    }
+    for (let i = 0; i < pathSegments.length; i++) {
+      var subPropertyKey = pathSegments[i];
+      if (!(subPropertyKey in document)) {
+        return undefined;
+      }
+      document = document[subPropertyKey];
+    }
+    return document;
+  }
+  /**
+   * Apply a single JSON Patch Operation on a JSON document.
+   * Returns {newDocument, result} of the operation.
+   *
+   * @param document The document to patch
+   * @param operation The operation to apply
+   * @param validate Whether to validate the operation
+   * @param mutateDocument Whether to mutate the original document or clone it before applying
+   * @return `{newDocument, result}` after the operation
    */
-  export function apply(tree: any, patches: Patch<any>[], validate?: boolean): any[] {
-    var results = new Array(patches.length)
-      , p = 0
-      , plen = patches.length
-      , patch
-      , key;
-    while (p < plen) {
-      patch = patches[p];
-      p++;
-      // Find the object
-      var path = patch.path || "";
-      var keys = path.split('/');
-      var obj = tree;
-      var t = 1; //skip empty element - http://jsperf.com/to-shift-or-not-to-shift
-      var len = keys.length;
-      var existingPathFragment = undefined;
-
-      while (true) {
+  export function applyOperation<T>(document: T, operation: Operation): OperationResult;
+
+  /**
+   * Apply a single JSON Patch Operation on a JSON document.
+   * Returns the {newDocument, result} of the operation.
+   *
+   * @param document The document to patch
+   * @param operation The operation to apply
+   * @param validate Whether to validate the operation
+   * @param mutateDocument Whether to mutate the original document or clone it before applying
+   * @return `{newDocument, result}` after the operation
+   */
+  export function applyOperation<T>(document: any, operation: Operation, validate: boolean, mutateDocument: boolean): OperationResult;
+  export function applyOperation<T>(document: any, operation: Operation, validate = false, mutateDocument = true): OperationResult {
+    if (validate) {
+      this.validator(operation, 0);
```

Comment:
wait, so if you just call `applyValidation(document, op, true)` the function will throw? Why does the user have to provide this as a `this` binding, instead of an optional parameter?